db/consistency_level.cc

```diff
@@ -23,8 +24,64 @@
 #include "locator/abstract_replication_strategy.hh"
 #include "locator/network_topology_strategy.hh"
 #include "heat_load_balance.hh"
+#include "tools/build_info.hh"
 
 namespace db {
+namespace {
+
+// Check the effective replication map consistency:
+// we have an inconsistent effective replication map in case we have a data center
+// with the number of replica nodes higher than it's replication factor.
+void validate_replica_dc_replication_factor(const locator::effective_replication_map& erm,
+                                            const inet_address_vector_replica_set& replicas,
+                                            const consistency_level cl) {
+    // Skip for release builds.
+    if constexpr (tools::build_info::is_release_build()) {
+        return;
+    }
+
+    // Applies only for Local Quorum consistency level.
+    if (cl != consistency_level::LOCAL_QUORUM) {
+        return;
+    }
+
+    const auto& topology = erm.get_topology();
+    const auto& replication_strategy = erm.get_replication_strategy();
+    // Applies only to Network Topology Replication Strategy.
+    if (replication_strategy.get_type() != locator::replication_strategy_type::network_topology) {
+        return;
+    }
+
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-static-cast-downcast)
+    const auto* network_topology_strategy = static_cast<const locator::network_topology_strategy*>(&replication_strategy);
+
+    struct rf_node_count {
+        size_t replication_factor{0};
+        size_t node_count{0};
+    };
+
+    boost::unordered_flat_map<sstring, rf_node_count> data_centers_replication_factor_;
+    for (const auto& node : replicas) {
+        auto res = data_centers_replication_factor_.emplace(topology.get_datacenter(node), rf_node_count{0, 0});
+        if (res.second) {
+            // For new item add replication factor.
+            res.first->second.replication_factor = network_topology_strategy->get_replication_factor(res.first->first);
+        }
+        ++res.first->second.node_count;
+    }
+
+    std::ranges::for_each(data_centers_replication_factor_, [](const auto& item) {
+        if (item.second.replication_factor < item.second.node_count) {
+            on_internal_error(cl_logger,
+                              seastar::format("ERM inconsistency, Datacenter [{}] has higher count of replicas [{}] than it's replication factor [{}]",
```

Comment:
But this check only look at replicas already filtered for query (i.e. CL taken into account), or total number of read replica candidates (before taking CL into account)? Unclear from the function comment

Reply:
Modified the message